playerbot/BroadcastHelper.cpp

```diff
@@ -0,0 +1,922 @@
+    if (locale == 255)
+        locale = static_cast<int32>(LocaleConstant::LOCALE_enUS);
+    return locale;
+}
+
+bool BroadcastHelper::BroadcastTest(
+    PlayerbotAI* ai,
+    Player* bot
+)
+{
+    //return something to ignore the logic
+    return false;
+
+    //don't think bots rejoin channnels when changing areas?
+    //bot->GetPlayerbotMgr()->JoinChatChannels(bot);
+
+    std::map<std::string, std::string> placeholders;
+    placeholders["%rand1"] = std::to_string(urand(0, 1));
+    placeholders["%rand2"] = std::to_string(urand(0, 1));
+    placeholders["%rand3"] = std::to_string(urand(0, 1));
+
+    int32 rand = urand(0, 1);
+
+    if (rand == 1 && ai->SayToTrade(BOT_TEXT2("Posted to trade, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (ai->SayToGuildRecruitment(BOT_TEXT2("Posted to GuildRecruitment, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+
+    return ai->SayToTrade(BOT_TEXT2("Posted to trade, %rand1, %rand2, %rand3", placeholders));
+
+    //int32 rand = urand(1, 8);
+    if (rand == 1 && ai->SayToGuild(BOT_TEXT2("Posted to guild, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 2 && ai->SayToWorld(BOT_TEXT2("Posted to world, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 3 && ai->SayToGeneral(BOT_TEXT2("Posted to general, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 4 && ai->SayToTrade(BOT_TEXT2("Posted to trade, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 5 && ai->SayToLFG(BOT_TEXT2("Posted to LFG, %rand1, %rand2, %rand3", placeholders)))//
+        return true;
+    else if (rand == 6 && ai->SayToLocalDefense(BOT_TEXT2("Posted to LocalDefense, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 7 && ai->SayToWorldDefense(BOT_TEXT2("Posted to WorldDefense, %rand1, %rand2, %rand3", placeholders)))//+
+        return true;
+    else if (rand == 8 && ai->SayToGuildRecruitment(BOT_TEXT2("Posted to GuildRecruitment, %rand1, %rand2, %rand3", placeholders)))//
+        return true;
+
+    return false;
+}
+
+/**
+@param toChannels - map of (ToChannel, chance), where chance is in range 0-100 as uint32 (unless global chance is not 100%)
+
+@return true if said to the channel, false otherwise
+*/
+bool BroadcastHelper::BroadcastToChannelWithGlobalChance(PlayerbotAI* ai, std::string message, std::list<std::pair<ToChannel, uint32>> toChannels)
+{
+    if (message.empty()) {
+        return false;
+    }
+
+    for (const auto& pair : toChannels) {
+        uint32 chance = pair.second;
+        switch (pair.first)
+        {
+            case TO_GUILD:
+            {
+                if (urand(1, 100) <= chance
+                    && urand(1, sPlayerbotAIConfig.broadcastChanceMaxValue) <= sPlayerbotAIConfig.broadcastToGuildGlobalChance
```

Comment:
he meant the documentation itslef

1-100 would be TOO low, 1-30000 is optimal

Reply:
> he meant the documentation (comments) itslef, which I will fix
> 
> 1-100 would be TOO low, 1-30000 is optimal

Why not use floating point for the user settings? e.g.
BroadcastToGuildGlobalChance  = 0.1 for 1/1000 chance
BroadcastToGuildGlobalChance = 0.00333 for ~ 1/30000 chance

Reply:
I have changed comments in .dist file and replaced BroadcastChanceMaxValue with AiPlayerbot.EnableBroadcasts , that should be more user-friendly

but I have used uint32 so that it would consume a bit less memory and no float calculation